components/script/dom/htmlinputelement.rs

```diff
@@ -292,12 +316,71 @@ impl HTMLInputElementMethods for HTMLInputElement {
 
     // https://html.spec.whatwg.org/multipage/#dom-input-value
     fn Value(&self) -> DOMString {
-        self.textinput.borrow().get_content()
+
+        match self.get_value_mode() {
+            // On getting, it must return the current value
+            // (https://html.spec.whatwg.org/multipage/forms.html#concept-fe-value)
+            // of the element.
+            ValueMode::Value => self.textinput.borrow().get_content(),
+
+            // On getting, if the element has a value attribute, it
+            // must return that attribute's value; otherwise, it must
+            // return the string "on" (for default/on mode)
+            ValueMode::DefaultOn => {
+                let content = self.textinput.borrow().get_content();
```

Comment:
This should instead upcast itself to an `Element` and grab the content attribute `value` from it.